crates/energy/src/graph.rs

```diff
@@ -0,0 +1,259 @@
+    }
+
+    fn remove_matching(&mut self, entity: NearbyEntity) {
+        self.receivers.retain(|&e| e != entity);
+        self.producers.retain(|&e| e != entity);
+    }
+}
+
+#[derive(Component, Default, Debug, Clone)]
+pub struct NearbyUnits {
+    units: Nearby,
+    last_pos: Option<Vec2>,
+}
+
+impl NearbyUnits {
+    pub fn units(&self) -> TinyVec<[NearbyEntity; 64]> {
+        self.units.both()
+    }
+}
+
+fn setup(mut commands: Commands) {
+    commands.insert_resource(PowerGrid::default());
+}
+
+fn clean_up(mut commands: Commands) {
+    commands.remove_resource::<PowerGrid>();
+}
+
+/// This system spawns Energy Producers and Energy Receivers and nearby units.
+fn spawn_graph_components(
+    mut commands: Commands,
+    newly_spawned_units: Query<Entity, Added<Active>>,
+) {
+    for entity in newly_spawned_units.iter() {
+        commands
+            .entity(entity)
+            .insert((EnergyReceiver, NearbyUnits::default()));
+    }
+}
+
+pub fn update_nearby_recv(
+    spacial_index_producer: SpatialQuery<Entity, With<EnergyProducer>>,
+    spacial_index_receiver: SpatialQuery<Entity, With<EnergyReceiver>>,
+    mut nearby_units: Query<(Entity, &mut NearbyUnits, &Transform), Changed<Transform>>,
+) {
+    nearby_units
+        .par_iter_mut()
+        .for_each_mut(|(entity, mut nearby_units, transform)| {
+            if let Some(last_pos) = nearby_units.last_pos {
+                if transform.translation.to_flat().distance_squared(last_pos) < 0.5 {
+                    return;
+                }
+            }
+            nearby_units.last_pos = Some(transform.translation.to_flat());
+
+            let aabb = &Aabb::new(
+                Point::from(transform.translation - Vec3::splat(MAX_DISTANCE)),
+                Point::from(transform.translation + Vec3::splat(MAX_DISTANCE)),
+            );
+
+            let producers = spacial_index_producer.query_aabb(aabb, Some(entity));
+
+            let receivers = spacial_index_receiver.query_aabb(aabb, Some(entity));
+
+            nearby_units.units.clear();
+
+            nearby_units
+                .units
+                .producers
+                .extend(producers.map(|entity| entity.into()));
+            nearby_units
+                .units
+                .receivers
+                .extend(receivers.map(|entity| entity.into()));
+        });
+}
+
+fn update_graph(
+    mut power_grid: ResMut<PowerGrid>,
+    nearby_units: Query<(Entity, &NearbyUnits), Changed<NearbyUnits>>,
+) {
+    for (entity, nearby_units) in nearby_units.iter() {
+        let mut edges_to_remove = HashSet::new();
+
+        edges_to_remove.extend(
+            &mut power_grid
+                .graph
+                .edges(entity)
+                .map(|edge| (edge.source(), edge.target())),
+        );
+
+        let mut edges_to_add = vec![];
+
+        for nearby_entity in nearby_units.units.both() {
+            edges_to_add.push((entity, nearby_entity.into()));
```

Comment:
No need to create `edges_to_add`, it is used only once so you can iterate `nearby_units.units.both()` directly.